src/App.tsx

```diff
@@ -1,26 +1,214 @@
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch(() => setError(Error.UnableToLoadAll));
+  }, []);
+
+  const filteredTodos = useMemo(() => {
+    return filterTodos(todos, filter);
+  }, [todos, filter, filterTodos]);
+
+  const activeTodosCount = useMemo(() => {
+    return todos.filter(todo => todo.completed !== true).length;
+  }, [todos]);
+
+  const isCompletedTodos = useMemo(() => {
+    return todos.filter(todo => todo.completed === true).length > 0;
+  }, [todos]);
+
+  const addNewTodo = useCallback((title: string) => {
+    if (!title.trim()) {
+      setError(Error.NoTitle);
+
+      return;
+    }
+
+    const temp = {
+      title,
+      id: 0,
+      completed: false,
+      userId: USER_ID,
+    };
+
+    setTempTodo(temp);
+    setError(null);
+
+    addTodos(temp)
+      .then(res => {
+        setTodos(prev => [...prev, res]);
+      })
+      .catch(() => setError(Error.UnableToAdd))
+      .finally(() => {
+        setTempTodo(null);
+      });
+  }, []);
+
+  const deleteCurrentTodo = useCallback((id: number) => {
+    setLoadingTodoIds(prev => [...prev, id]);
+    deleteTodos(id)
+      .then(() => {
+        setTodos(prev => prev.filter(todo => todo.id !== id));
+      })
+      .catch(() => setError(Error.UnableToDelete))
+      .finally(() => {
+        setLoadingTodoIds([]);
+      });
+  }, []);
+
+  const deleteCompletedTodos = useCallback(() => {
+    const completedTodoIds = todos
+      .filter(todo => todo.completed)
+      .map(todo => todo.id);
+
+    setLoadingTodoIds(completedTodoIds);
+
+    Promise.all(completedTodoIds.map(id => deleteTodos(id)))
+      .then(() => {
+        setTodos(prev =>
+          prev.filter(todo => !completedTodoIds.includes(todo.id)),
+        );
+      })
+      .catch(() => setError(Error.UnableToDelete))
+      .finally(() => {
+        setLoadingTodoIds([]);
+      });
+  }, [todos]);
+
+  const allCompleted = todos.every(todo => todo.completed);
+
+  const toggleAllTodos = () => {
+    if (allCompleted) {
+      const loadingTodosIds = todos.map(todo => todo.id);
+
+      setLoadingTodoIds(loadingTodosIds);
+
+      const activeTodos = todos.map(todo => ({
+        ...todo,
+        completed: false,
+      }));
+
+      const patchPromises = todos.map(todo =>
+        patchTodo(todo.id, { ...todo, completed: false }),
+      );
+
+      Promise.all(patchPromises)
+        .then(() => {
+          setTodos(activeTodos);
+        })
+        .catch(() => setError(Error.UnableToUpdate))
+        .finally(() => setLoadingTodoIds([]));
+
+      return;
+    }
+
+    const loadingTodosIds = todos
+      .filter(todo => !todo.completed)
+      .map(todo => todo.id);
+
+    setLoadingTodoIds(loadingTodosIds);
+
+    const completedTodos = todos.map(todo => ({
+      ...todo,
+      completed: true,
+    }));
+
+    const patchPromises = todos.map(todo =>
+      patchTodo(todo.id, { ...todo, completed: true }),
+    );
```

Comment:
This logic should be reworked to not affect for todos which were already toggled.